### PR TITLE
#217 side chain flexion probabilities

### PR DESCRIPTION
**Pull request by ssepeq**
The new "flexion selection" feature correctly predicts that nonanoic acid is a non-agonist for OR51E2. The tail end of the ligand stays in the binding pocket, without moving Phe155, and the carboxylic acid makes contact with Arg262 almost right away in node 1 of the sweep, obviating receptor activation. (It's not known whether nonanoic acid is an antagonist, but this pattern of binding suggest antagonism would be likely.)

Running it on the server now to see what kind of accuracy it gets.

### CODE REVIEW

**Comment by ssepeq** [comment]
🟥🟩...

TAAR test only took 235 seconds! Much faster now with fewer flexions. Incorrectly, it reports His107 as having a moderate repulsion with the ligand, when in fact there appear to be about 5A between them, besides the fact that this side chain would not be expected to have a negative charge in the configuration shown.

**Comment by ssepeq** [comment]
The calculations on the server are producing some strange effects. Just checked the dock for propionic acid in 51E1 (should be an agonist) and there are no path nodes beyond node zero.

**Comment by primaryodors** [comment]
The side chain flexions appear to be working as they should. Helix rotations are beyond the scope of this PR.

Should get this branch ready for merge (make sure it passes the TAAR and 1A1 tests) and then tackle the helix rotations in a new branch.

**Comment by ssepeq** [comment]
🟥🟩🟥 

@%$#.

**Comment by ssepeq** [comment]
🟥🟩🟩

File: predict/A51E2.pepd [hunk 1]
```diff
@@ -0,0 +1,327 @@
+# Pepteditor script for deriving active-state conformational changes from human β2 adrenergic receptor.
```
Comment (primaryodors) [inline predict/A51E2.pepd:1]: Description comment is wrong.

File: predict/A51E2.pepd [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+# Output the rotations.
+
+ECHO "# Active rotations from A51E2.pepd:"
+ECHO "$rotations ="
+ECHO "["
+
+LET %i = 1
+_looprot:
+
+# Origin will be the active m point.
```
Comment (primaryodors) [inline predict/A51E2.pepd:274]: These comments are also wrong.

File: predict/A51E2.pepd [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+LET @xform.y /= 2
+LET @xform.z /= 2
+
+LET @cen = @ac[%i] + @an[%i]
+
+LET @cen.x /= 2
+LET @cen.y /= 2
+LET @cen.z /= 2
+
+LET @point = @ic[%i] + @delta_n
```
Comment (primaryodors) [inline predict/A51E2.pepd:292]: Should this be @ delta_n or @ xform?

File: predict/A51E2.pepd [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+LET %i ++
+
+IF %i <= 7 GOTO _looprot
+ECHO "];"
+
+
+
+
+# DUMP
```
Comment (primaryodors) [inline predict/A51E2.pepd:324]: This is no longer in use and should be removed.

File: predict/analyze_emrot.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+echo "\n";
+
+if (!$total) echo "Not enough data processed.\n\n";
+else echo "$correct correct predictions out of $total total ($agonists agonists / $nonagonists other), $fp false positives / $fn false negatives, "
+  . round(100.0 * $correct / $total, 2) . "% accuracy.\n\n";
+
+
+
```
Comment (primaryodors) [inline predict/analyze_emrot.php:95]: Remove extra blank lines from end of file.

File: predict/method_emrots.php [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+$dock_metals = false;
+$bias_by_energy = true;
+
+chdir(__DIR__);
+require("methods_common.php");
+chdir(__DIR__);
+
+prepare_outputs();
+
+// $outfname = str_replace(".dock", "_emrot.dock", $outfname);
```
Comment (primaryodors) [inline predict/method_emrots.php:23]: Let's remove this comment and the comment block below.

File: predict/methods_common.php [hunk 7]
```diff
@@ -241,6 +249,14 @@ function process_dock($metrics_prefix = "", $noclobber = false)
             continue;
         }
 
+        if ($pose && $node>=0 && substr($ln, 0, 15) == "Protein clashes")
+        {
+            // echo "$ln\n";
```
Comment (primaryodors) [inline predict/methods_common.php:254]: Let's remove this debug line.

File: predict/methods_common.php [hunk 8]
```diff
@@ -241,6 +249,14 @@ function process_dock($metrics_prefix = "", $noclobber = false)
...
+            $pclash[$pose][$node] = floatval(explode(": ", $ln)[1]);
+            $dosce = false;
+            continue;
+        }
+
         if ($pose && $node>=0 && strpos($ln, " active theta: "))
         {
             // echo "$ln\n";
```
Comment (primaryodors) [inline predict/methods_common.php:262]: And this one.

File: src/classes/aminoacid.cpp [hunk 9]
```diff
@@ -1474,12 +1518,8 @@ bool AminoAcid::can_reach(AminoAcid* other) const
         return false;
     }
 
-    //cout << aadef->_3let << residue_no << " vs " << other->aadef->_3let << other->residue_no;
```
Comment (primaryodors) [inline src/classes/aminoacid.cpp:1477]: Should remove this debug line and the one below it, or place them uncommented inside a compiler switch defined in constants.h.

File: src/classes/atom.cpp [hunk 10]
```diff
@@ -1257,6 +1257,12 @@ bool Bond::rotate(float theta, bool allow_backbone, bool skip_inverse_check)
     }
     // cout << "Rotate " << atom->name << cardinality_printable(cardinality) << btom->name << endl;
 
+    if (atom->residue && btom->residue)
```
Comment (primaryodors) [inline src/classes/atom.cpp:1260]: Should remove this debug trap.

File: src/classes/molecule.cpp [hunk 11]
```diff
@@ -2709,6 +2709,11 @@ void Molecule::multimol_conform(Molecule** mm, Molecule** bkg, Molecule** ac, in
             float lreversal = -pow(0.1, 2.0/iters);
             float accel = 1.01;
 
+
+            #if _dbg_multiflex
+            // cout << iter << ":" << mm[i]->name << " " << mm[i]->movability << endl;
```
Comment (primaryodors) [inline src/classes/molecule.cpp:2714]: Should either uncomment this or remove entirely.

File: src/primarydock.cpp [hunk 12]
```diff
@@ -1660,15 +1695,7 @@ void prepare_initb()
             );
             delete[] delete_me;
         }
-
-        /*
```
Comment (primaryodors) [inline src/primarydock.cpp:1664]: Should remove this comment block.

**Review by primaryodors — CHANGES_REQUESTED** [review]

File: src/classes/aminoacid.cpp (hunk 9 above)
Comment (ssepeq) [reply src/classes/aminoacid.cpp:1477]: These were already removed. 😕

File: src/primarydock.cpp (hunk 12 above)
Comment (ssepeq) [reply src/primarydock.cpp:1664]: Consider it done. 👍🏼

**Review by primaryodors — APPROVED** [review]

**Pull request merged**